pkg/routeagent-driver/handlers/kubeproxy-iptables/endpoint_handler.go

```diff
@@ -0,0 +1,169 @@
+		err = kp.createVxLANInterface(VxInterfaceWorker, localClusterGwNodeIP)
+		if err != nil {
+			klog.Fatalf("Unable to create VxLAN interface on non-GatewayNode (%s): %v", endpoint.Spec.Hostname, err)
+		}
+
+		kp.vxlanGwIP = &remoteVtepIP
+		err = kp.reconcileRoutes(remoteVtepIP)
+		if err != nil {
+			return fmt.Errorf("error while reconciling routes %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (kp *SyncHandler) LocalEndpointUpdated(endpoint *submV1.Endpoint) error {
+	return nil
+}
+
+func (kp *SyncHandler) LocalEndpointRemoved(endpoint *submV1.Endpoint) error {
+	kp.syncHandlerMutex.Lock()
+	defer kp.syncHandlerMutex.Unlock()
+	kp.isGatewayNode = false
+
+	if kp.vxlanDevice != nil {
+		err := kp.vxlanDevice.deleteVxLanIface()
+		kp.vxlanDevice = nil
+		kp.vxlanGwIP = nil
+		if err != nil {
+			return fmt.Errorf("Failed to delete the the vxlan interface on Endpoint removal: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (kp *SyncHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	if err := kp.overlappingSubnets(endpoint.Spec.Subnets); err != nil {
+		// Skip processing the endpoint when CIDRs overlap
+		return err
+	}
+
+	kp.syncHandlerMutex.Lock()
+	defer kp.syncHandlerMutex.Unlock()
+	for _, inputCidrBlock := range endpoint.Spec.Subnets {
+		if !kp.remoteSubnets.Contains(inputCidrBlock) {
+			kp.remoteSubnets.Add(inputCidrBlock)
+		}
+	}
+
+	if err := kp.updateRoutingRulesForInterClusterSupport(endpoint.Spec.Subnets, Add); err != nil {
+		klog.Errorf("updateRoutingRulesForInterClusterSupport returned error: %+v", err)
+		return err
+	}
+	// Add routes to the new endpoint on the GatewayNode.
+	kp.updateRoutingRulesForHostNetworkSupport(endpoint.Spec.Subnets, Add)
+	kp.updateIptableRulesForInterClusterTraffic(endpoint.Spec.Subnets, Add)
+
+	return nil
+}
+
+func (kp *SyncHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
+	return nil
+}
+
+func (kp *SyncHandler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
+	kp.syncHandlerMutex.Lock()
+	defer kp.syncHandlerMutex.Unlock()
+	for _, inputCidrBlock := range endpoint.Spec.Subnets {
+		if !kp.remoteSubnets.Contains(inputCidrBlock) {
+			kp.remoteSubnets.Delete(inputCidrBlock)
+		}
+	}
+	// TODO: Handle a remote endpoint removal use-case
+	//         - remove related iptable rules
+	if err := kp.updateRoutingRulesForInterClusterSupport(endpoint.Spec.Subnets, Delete); err != nil {
+		klog.Errorf("updateRoutingRulesForInterClusterSupport returned error: %+v", err)
+		return err
+	}
+
+	kp.updateRoutingRulesForHostNetworkSupport(endpoint.Spec.Subnets, Delete)
+	kp.updateIptableRulesForInterClusterTraffic(endpoint.Spec.Subnets, Delete)
+
+	return nil
+}
+
+func (kp *SyncHandler) overlappingSubnets(remoteSubnets []string) error {
+	// If the remoteSubnets [*] overlap with local cluster Pod/Service CIDRs we
+	// should not update the IPTable rules on the host, as it will disrupt the
+	// functionality of the local cluster. So, lets validate that subnets do not
```

Comment:
Ideally we may want to look also for the local worker nodes network CIDR at some point, but that's an existing problem today, not related to your new code.